Feature/create ioc

josera980 opened the pull request:
Creating IoC file. Used an Object data structure since we don't have to iterate when registering or getting data.

josera980 committed:
Creating IoC file.

reynoldmorel commented on src/ioc/Ioc.js:
```diff
@@ -0,0 +1,17 @@
+        this.services = {};
+    }
+
+    register(service, serviceName) {
```
Change the order, should be serviceName first and then service.

reynoldmorel replied:
Also specify the method is `public`

reynoldmorel replied:
Also please throw and error if `serviceName` is falsy

reynoldmorel commented on src/ioc/Ioc.js:
```diff
@@ -0,0 +1,17 @@
+
+    register(service, serviceName) {
+        if (this.services[serviceName]) {
+            console.log("Service already created.")
```
Throw an error here instead

reynoldmorel replied:
Also you won't need a `if / else` block if you do the above. I think just an if will `work`

reynoldmorel commented on src/ioc/Ioc.js:
```diff
@@ -0,0 +1,17 @@
+        }
+    }
+
+    getService(serviceName) {
```
specify the method is public. Also please throw and error if `serviceName` is falsy

reynoldmorel replied:
Some here I think you can also simplify this here

josera980 commented:
Changed order of service and serviceName.

Added a validation to check if `serviceName` is falsy.

I have the `if / else` block so I can throw different error messages, depending on the circumstances.

Also, methods are public by default if we do not specify it or if we do not add a prefix.

reynoldmorel commented on src/ioc/Ioc.js:
```diff
@@ -0,0 +1,25 @@
+    register(serviceName, service) {
+        if (this.services[serviceName]) {
+            throw new Error("Service already created.");
+        } else if (typeof serviceName === "string" && serviceName.length > 0) {
```
You can also simplify like this:

```
 if (!serviceName) {
    throw new Error("Please verify service name.");
}

 if (this.services[serviceName]) {
    throw new Error("Service already created.");
}

this.services[serviceName] = service;
```

josera980 replied:
Done. Code simplified.

josera980 commented:
About the public methods, we can find information about it here:
[https://www.javascripttutorial.net/javascript-private-methods/#:~:text=By%20default%2C%20members%20of%20a,methods%2C%20and%20getter%2Fsetters.](url)

`By default, members of a [class](https://www.javascripttutorial.net/es6/javascript-class/) are public. ES2020 introduced the private members that include [private fields](https://www.javascripttutorial.net/javascript-private-fields/) and methods. 

To make a public method private, you prefix its name with a hash #. JavaScript allows you to define private methods for instance methods, [static methods](https://www.javascripttutorial.net/es6/javascript-static-method/), and [getter/setters](https://www.javascripttutorial.net/es6/javascript-getters-and-setters/).`

reynoldmorel commented on src/ioc/Ioc.js:
```diff
@@ -0,0 +1,35 @@
+let services = {};
```
You don't need to put this here. You can move it inside the constructor as it was before

josera980 replied:
But if I have it inside, would I be able to modify it (register services) having the `Object.freeze()` on the instance?

josera980 replied:
It works having it inside the constructor. Changed it now.

josera980 committed:
Added linter and formatter

josera980 commented:
New commit adding linter (eslint) and formatter (prettier).

josera980 committed:
adding jest and unit test to ioc

josera980 commented:
New commit adding jest and a unit test to Ioc.

reynoldmorel reviewed (changes requested):
1. Why are all entities been created again?
2. Did you notice a bunch of files without EOF `new liine`?
3. Just curious... Why are you adding babel?
4. Do not commit coverage folder, use gitignore
5. Could you re-think your tests? could you read about unit testing and integration testing? The tests for IoC which are adding is not very good
6.
